testutils/resourcebuilder/machine/v1beta1/machineset.go

```diff
@@ -27,55 +28,108 @@ const (
+	defaultReadyReplicas          = int32(0)
+	defaultReplicasStatus         = int32(0)
+	defaultSynchronizedGeneration = int64(0)
+)
+
 // MachineSet creates a new machineSet builder.
 func MachineSet() MachineSetBuilder {
 	return MachineSetBuilder{}
 }
 
 // MachineSetBuilder is used to build out a machineSet object.
 type MachineSetBuilder struct {
-	generateName        string
-	name                string
-	namespace           string
-	replicas            int32
-	labels              map[string]string
-	creationTimestamp   metav1.Time
-	providerSpecBuilder resourcebuilder.RawExtensionBuilder
+	authoritativeAPI           *machinev1beta1.MachineAuthority
+	creationTimestamp          metav1.Time
+	deletePolicy               *string
+	deletionTimestamp          *metav1.Time
+	generateName               string
+	lifecycleHooks             *machinev1beta1.LifecycleHooks
+	machineSetAnnotations      map[string]string
+	machineSetLabels           map[string]string
+	machineSpec                *machinev1beta1.MachineSpec
+	machineSpecObjectMeta      *machinev1beta1.ObjectMeta
+	machineTemplateAnnotations map[string]string
+	machineTemplateLabels      map[string]string
+	minReadySeconds            *int32
+	name                       string
+	namespace                  string
+	providerSpec               *machinev1beta1.ProviderSpec
+	providerSpecBuilder        *resourcebuilder.RawExtensionBuilder
+	replicas                   *int32
+	taints                     []corev1.Taint
 
 	// status fields
-	errorMessage *string
+	authoritativeAPIStatus *machinev1beta1.MachineAuthority
+	availableReplicas      *int32
+	conditions             []machinev1beta1.Condition
+	errorMessage           *string
+	errorReason            *machinev1beta1.MachineSetStatusError
+	fullyLabeledReplicas   *int32
+	observedGeneration     *int64
+	readyReplicas          *int32
+	replicasStatus         *int32
+	synchronizedGeneration *int64
 }
 
 // Build builds a new machineSet based on the configuration provided.
 func (m MachineSetBuilder) Build() *machinev1beta1.MachineSet {
 	machineSet := &machinev1beta1.MachineSet{
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName:      m.generateName,
+			Annotations:       m.machineSetAnnotations,
 			CreationTimestamp: m.creationTimestamp,
+			DeletionTimestamp: m.deletionTimestamp,
+			GenerateName:      m.generateName,
+			Labels:            m.machineSetLabels,
 			Name:              m.name,
 			Namespace:         m.namespace,
-			Labels:            m.labels,
 		},
 		Spec: machinev1beta1.MachineSetSpec{
-			Replicas: &m.replicas,
+			AuthoritativeAPI: resourcebuilder.Coalesce(m.authoritativeAPI, defaultAuthoritativeAPI),
+			DeletePolicy:     resourcebuilder.Coalesce(m.deletePolicy, defaultDeletePolicy),
+			MinReadySeconds:  resourcebuilder.Coalesce(m.minReadySeconds, defaultMinReadySeconds),
+			Replicas:         m.replicas,
 			Selector: metav1.LabelSelector{
-				MatchLabels: m.labels,
+				MatchLabels: m.machineTemplateLabels,
```

Comment:
Is there an occasion you can think of where we might want to try an alternative selector?

Reply:
The selector should be matchng the labels in the Template.ObjectMeta.Labels below, so this and that one should match.
Although the MachineSet.ObjectMeta.Labels, are different right?

Reply:
Yes, MachineSet.ObjectMeta.Labels are different.

My thinking here was that perhaps we might want to use this builder to provide a case where the selector uses match expressions, or, doesn't match the labels on the template.

I think having a builder that allows providing an entire selector, and then defaulting to what you've included here may be a more flexible solution

Reply:
Makes sense. Extended the builder to accept a custom selector with `WithMachineSetSpecSelector()`, this may then differ from the machineSet.Spec.Template.ObjectMeta.Labels, which can separately be set with `WithMachineTemplateLabels()`.